### PR TITLE
Add PubSub integration guide to docs

### LINKED ISSUE #2454: Document pub/sub integrations
#### Summary

Document application pub/sub integrations

#### Why do we need this ?

To complete the application data integrations

#### What is already there? What do you see now?

MQTT Server and HTTP Webhooks

#### What is missing? What do you want to see?

PubSub integrations

#### How do you propose to document this?

Add section like MQTT Server and Webhooks

It's good to mention also that we have two sorts of MQTT integrations; we have a builtin server (MQTT Server), and we have a client (via PubSub)

#### Can you do this yourself and submit a Pull Request?

Can review

### PR DESCRIPTION
**Pull request by benolayinka**
#### Summary

Closes #2454 

#### Changes

- Add PubSub integration guide to docs

#### Notes for Reviewers

I couldn't get mbed-simulator to work, so next week I will run thru this guide using a real gateway connected to my dev stack. Pending self-review.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
- [ ] Testing: The changes are covered with unit tests. The changes are tested manually as well.
- [x] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: doc/content/integrations/pubsub/message-types/_index.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+This will also result in the following message being published by the Application Server
+
+```bash
+# Received on [downlink-queued-subtopic] : 
+'{"end_device_ids":{"device_id":"dev1","application_ids":{"application_id":"app1"}},"correlation_ids":["as:downlink:01DAVNFG65NAMC5DMX0GFJ8CSK"],"downlink_queued":{"f_port":15,"frm_payload":"vu8=","priority":"NORMAL","correlation_ids":["as:downlink:01DAVNFG65NAMC5DMX0GFJ8CSK"]}}'
+```
+
+## API Key
+>No API key is needed for PubSub messaging. All messages are assumed to be authorized.
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/message-types/_index.md:36]: This is not really a section this way; I'd suggest moving this up and remove the heading

File: doc/content/integrations/pubsub/message-types/_index.md [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+## Message Format
+
+JSON messages sent or received by the Application Server are defined in the [Data Formats]({{< ref "/integrations/data-formats" >}}) section.
+
+## Scheduling Downlinks
+
+You can schedule downlink messages by publishing to a topic. For example, using a `nats-pub` client connected to your NATS server:
+
+```bash
+$ nats-pub push-subtopic '{"end_device_ids":{"device_id":"dev1","application_ids":{"application_id":"app1"}},"downlinks":[{"f_port":15,"frm_payload":"vu8=","priority":"NORMAL"}]}'
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/message-types/_index.md:23]: This section is about message types, but mentions NATS instructions here. I think that is confusing. What is the goal of this section, other than saying that it follows the data format used everywhere?

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+title: "Configure MQTT (Client)"
+description: ""
+weight: 1
+---
+
+This section explains how to configure a PubSub Integration using the built in MQTT client and an external MQTT server.
+
+<!--more-->
+
+Creating a PubSub integration with the built in MQTT Client requires you to have an MQTT server running with an available endpoint. [Mosquitto](https://mosquitto.org/) is a popular open source MQTT Server you may use, but {{% tts %}} MQTT client should work with any MQTT server.
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:11]: - client (lowercase)
- "should work" => works

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+In your application select the **Pub-subs** submenu from the **Integrations** side menu. Clicking on the **+ Add pub-sub** button will open the PubSub creation screen.
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:13]: "PubSubs" right? If it says "Pub-subs" and "pub-sub" we have an inconsistency we should fix there

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+Give your PubSub an ID, choose a format, and enter a **Base Topic** to subscribe or publish to.
+
+Select **MQTT** as provider.
+
+{{< figure src="pubsub-id.png" alt="Pubsub creation screen" >}}
+
+If using Client Certificates, enable **Secure**. You will be asked to upload the **Server root CA certificate**, the **Client certificate**, and the **Client private key.**
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:21]: "Root CA certificate"

Inconsistent casing "Base Topic" (title case) vs "Client private key" (sentence case). Is this literally how it is in the Console?

File: doc/content/integrations/pubsub/nats/_index.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+weight: 1
+---
+
+This section explains how to configure a PubSub Integration using an external NATS server.
+
+<!--more-->
+
+Creating a PubSub integration with NATS requires you to have a NATS server running with an available endpoint. See the [NATS Documentation](https://docs.nats.io/) for more information about configuring a NATS server.
+
+In your application select the **Pub-subs** submenu from the **Integrations** side menu. Clicking on the **+ Add pub-sub** button will open the PubSub creation screen.
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/nats/_index.md:13]: See previous comments

File: doc/content/integrations/pubsub/mqtt-client/_index.md (hunk 5 above)
Comment (benolayinka) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:21]: This is how it is in the console. Should they be title case?

Screenshot 2020-05-12 at 17 36 07
Screenshot 2020-05-12 at 17 36 14
Screenshot 2020-05-12 at 17 37 08

File: doc/content/integrations/pubsub/mqtt-client/_index.md (hunk 4 above)
Comment (benolayinka) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:13]: true. It's most commonly written **Pub/Sub** in the wild - should we go with that?

**Comment by benolayinka** [comment]
Blocked on #2516 - mqtt url scheme

File: doc/content/integrations/pubsub/mqtt-client/_index.md (hunk 5 above)
Comment (kschiffer) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:21]: We moved to sentence case in the Console recently, as per the design guidelines. If there is title case somewhere then that needs to be fixed.

File: doc/content/integrations/pubsub/mqtt-client/_index.md (hunk 4 above)
Comment (kschiffer) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:13]: Yeah had some problems finding a conventional way to write it. Maybe Pub/Subs is indeed better. I'll make a PR to change this.

File: doc/content/integrations/pubsub/message-types/_index.md [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+- Uplink Message
+- Join Accept
+- Downlink Ack
+- Downlink Nack
+- Downlink Sent
+- Downlink Failed
+- Downlink Queued
+- Location Solved
+- Downlink Push
+- Downlink Replace
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/message-types/_index.md:16]: Please split our publish vs subscribe

Also add "Service Data" (coming in #2541)

File: doc/content/integrations/pubsub/message-types/_index.md [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+ Enabling event messaging also allows you to manually configure a **Sub Topic** for that event. If no Sub Topic is specified, events will be published to the configured **Base Topic**.
+
+<!--more-->
+
+>Sub Topics should be specified for at least **Downlink Push** and **Downlink Replace** in order to schedule messages by publishing to these topics. If Downlink Push **AND** Downlink Replace are enabled, but have **no Sub Topic**, messages published to **Base Topic** may be scheduled as either Downlink Push **OR** Downlink Replace. This is an undesirable race condition.
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/message-types/_index.md:22]: OK so what should the user know here really?

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+
+This section explains how to configure a Pub/Sub Integration using the built in MQTT client.
+
+<!--more-->
+
+Creating a Pub/Sub Integration with the built in MQTT client requires you to have an MQTT server running with an available endpoint. [Mosquitto](https://mosquitto.org/) is a popular open source MQTT Server you may use, but {{% tts %}} MQTT client works with any MQTT server.
+
+>No API key is needed for Pub/Sub messaging. All messages received by the Application Server are assumed to be authorized, so be sure to configure appropriate security on your MQTT server.
+
+In your application select the **Pub-subs** submenu from the **Integrations** side menu. Clicking on the **+ Add pub-sub** button will open the Pub/Sub creation screen.
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:15]: As mentioned in earlier review; it is exactly **PubSubs** and **PubSub**

Other styles; pub-sub, Pub/Sub, etc are wrong

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+Give your Pub/Sub an **ID**, choose a **Message Format**, and enter a **Base Topic** to subscribe or publish to.
+
+Select **MQTT** as provider.
+
+{{< figure src="pubsub-id.png" alt="Pub/Sub creation screen" >}}
+
+If using Client Certificates, enable **Secure**. You will be asked to upload the **Server Root CA Certificate**, the **Client Certificate**, and the **Client Private Key.**
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:23]: "Client Certificates" => "TLS mutual authentication"

I see that **Client Private Key** has sentence casing in the Console; can you fix that as part of this PR?

While you're at it, you might want to change it to "TLS Server Root CA Certificate", "TLS Client Certificate" and "TLS Private Key"

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+The Application Server can be configured to publish and subscribe to topics for individual events. See [Message Types]({{< relref "../message-types" >}}) for more information about events and topics.
+
+After enabling events, click the **Add pub-sub** button to enable the integration.
+
+## Subscribing to Upstream Traffic
+
+The Application Server publishes messages for any enabled events. For example, when a device sends an uplink, an `uplink` message is published. To view these using a `mosquitto_sub` client connected to your MQTT server:
+
+```bash
+$ mosquitto_sub -t '#' -v
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:44]: Specify host name, even if it's localhost

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+# base-topic/uplink {"end_device_ids":{"device_id":"dev1","application_ids":{"application_id":"app1"}},"received_at":"2020-05-12T12:23:07.087614Z","uplink_message":{"session_key_id":"AXIDznz4bnQqtW8T3NsIVg==","f_port":1,"f_cnt":327,"frm_payload":"AQ=="}}
+```
+
+## Scheduling Downlinks
+
+You can schedule downlink messages by publishing to a topic. For example, using a `mosquitto_pub` client connected to your MQTT server:
+
+```bash
+$ mosquitto_pub -t 'base-topic/push-subtopic' -m '{"end_device_ids":{"device_id":"dev1","application_ids":{"application_id":"app1"}},"downlinks":[{"f_port":1,"frm_payload":"AA==","priority":"NORMAL"}]}'
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:53]: Idem. People may have an external MQTT server running, not necessarily on localhost

File: doc/content/integrations/pubsub/nats/_index.md [hunk 13]
```diff
@@ -0,0 +1,62 @@
+---
+title: "NATS Client"
+description: ""
+weight: 1
+---
+
+This section explains how to configure a Pub/Sub Integration using the built in NATS client.
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/nats/_index.md:7]: Idem about casing

File: doc/content/integrations/pubsub/nats/_index.md [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+The Application Server can be configured to publish and subscribe to topics for individual events. See [Message Types]({{< relref "../message-types" >}}) for more information about events and topics.
+
+After enabling events, click the **Add pub-sub** button to enable the integration.
+
+## Subscribing to Upstream Traffic
+
+The Application Server publishes messages for any enabled events. For example, when a device sends an uplink, an `uplink` message is published. To view these using a `nats-sub` client connected to your NATS server:
+
+```bash
+$ nats-sub '>'
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/nats/_index.md:42]: Idem also about host name

File: doc/content/integrations/pubsub/nats/_index.md [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+# Listening on [>]
+# [#1] Received on [base-topic.uplink-subtopic] : '{"end_device_ids":{"device_id":"dev1","application_ids":{"application_id":"app1"}, "received_at":"2020-05-12T10:12:42.063941Z","uplink_message":{"session_key_id":"AXIDznz4bnQqtW8T3NsIVg==","f_port":1,"f_cnt":102,"frm_payload":"AQ=="}]}'
+```
+
+## Scheduling Downlinks
+
+You can schedule downlink messages by publishing to a topic. For example, using a `nats-pub` client connected to your NATS server:
+
+```bash
+$ nats-pub base-topic.push-subtopic '{"end_device_ids":{"device_id":"dev1","application_ids":{"application_id":"app1"}},"downlinks":[{"f_port":1,"frm_payload":"AA==","priority":"NORMAL"}]}'
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/nats/_index.md:52]: Idem

File: doc/content/integrations/pubsub/mqtt-client/_index.md (hunk 10 above)
Comment (benolayinka) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:23]: Console labels are all sentence casing per design guidelines https://github.com/TheThingsIndustries/lorawan-stack/issues/2109#issuecomment-627690981

Comment (johanstokking) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:23]: They are title case, except this one:

Screenshot 2020-05-14 at 11 59 03

Comment (benolayinka) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:23]: hmm maybe enterprise hasn't been updated? open source looks like this

Screenshot 2020-05-12 at 17 36 07
Screenshot 2020-05-12 at 17 36 14
Screenshot 2020-05-12 at 17 37 08

Comment (johanstokking) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:23]: Ah ok, then it's fine!

Comment (johanstokking) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:23]: Still, change the names of the labels

File: doc/content/integrations/pubsub/message-types/_index.md (hunk 8 above)
Comment (benolayinka) [reply doc/content/integrations/pubsub/message-types/_index.md]: hard to reword but i gave it a shot. if you wanna use both push and replace subtopics, you should specify subtopics for both. if you don't, and publish messages on the base topic, the application server will randomly schedule them as push or replace.

File: doc/content/integrations/pubsub/message-types/_index.md (hunk 7 above)
Comment (johanstokking) [reply doc/content/integrations/pubsub/message-types/_index.md:16]: Please sort

Comment (johanstokking) [reply doc/content/integrations/pubsub/message-types/_index.md:16]: Also push and replace are queue operations, so "Downlink Queue Push" and "... Replace"

File: doc/content/integrations/pubsub/message-types/_index.md [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+- Downlink Failed
+- Downlink Queued
+- Location Solved
+- Downlink Push
+- Downlink Replace
+- Service Data
+
+The Application Server can be configured to **subscribe** to messages to schedule the following events:
+- Downlink Push
+- Downlink Receive
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/message-types/_index.md:21]: Idem

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+
+In your application select the **Pub/Subs** submenu from the **Integrations** side menu. Clicking on the **+ Add Pub/Sub** button will open the Pub/Sub creation screen.
+
+Give your Pub/Sub an **ID**, choose a **Message Format**, and enter a **Base Topic** to subscribe or publish to.
+
+Select **MQTT** as provider.
+
+{{< figure src="mqtt.png" alt="Pub/Sub creation screen" >}}
+
+If using Client Certificates, enable **Secure**. You will be asked to upload the **Root CA Certificate**, the **Client Certificate**, and the **Client Private Key.**
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:23]: - This is now sentence case, right?
- Can you change the labels as requested previously?

File: doc/content/integrations/pubsub/nats-client/_index.md [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+
+<!--more-->
+
+Creating a Pub/Sub integration with NATS requires you to have a NATS server running with an available endpoint. See the [NATS Documentation](https://docs.nats.io/) for more information about configuring a NATS server.
+
+>No API key is needed for Pub/Sub messaging. All messages received by the Application Server are assumed to be authorized, so be sure to configure appropriate security on your NATS server.
+
+In your application select the **Pub/Subs** submenu from the **Integrations** side menu. Clicking on the **+ Add Pub/Sub** button will open the Pub/Sub creation screen.
+
+Give your Pub/Sub an **ID**, choose a **Message Format**, and enter a **Base Topic** to subscribe or publish to.
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/nats-client/_index.md:17]: Idem about casing

Or am I confused now?

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 19]
```diff
@@ -0,0 +1,63 @@
+---
+title: "MQTT Client"
+description: ""
+weight: 1
+---
+
+This section explains how to configure a Pub/Sub integration using the built in MQTT client.
```
Comment (kschiffer) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:7]:
```suggestion
This section explains how to configure a Pub/Sub integration using the built-in MQTT client.
```

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+<!--more-->
+
+Creating a Pub/Sub integration with the built in MQTT client requires you to have an MQTT server running with an available endpoint. [Mosquitto](https://mosquitto.org/) is a popular open source MQTT Server you may use, but {{% tts %}} MQTT client works with any MQTT server.
+
+>No API key is needed for Pub/Sub messaging. All messages received by the Application Server are assumed to be authorized, so be sure to configure appropriate security on your MQTT server.
```
Comment (kschiffer) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:13]:
```suggestion
> No API key is needed for Pub/Sub messaging. All messages received by the Application Server are assumed to be authorized, so be sure to configure appropriate security on your MQTT server.
```

File: doc/content/integrations/pubsub/message-types/_index.md (hunk 16 above)
Comment (benolayinka) [reply doc/content/integrations/pubsub/message-types/_index.md:21]: Are the topics Title Case or Sentence case?

Comment (benolayinka) [reply doc/content/integrations/pubsub/message-types/_index.md:21]: and do you mean sort alphabetically? so

Publish:
- Downlink ack
- Downlink failed
- Downlink nack
- Downlink queued
- Downlink sent
- Join accept
- Location solved
- Service data
- Uplink message

Publish
- Downlink queue push
- Downlink queue receive

File: doc/content/integrations/pubsub/message-types/_index.md [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+The Application Server can be configured to **publish** messages for any of the following events:
+- Downlink ack
+- Downlink failed
+- Downlink nack
+- Downlink queued
+- Downlink sent
+- Join accept
+- Location solved
+- Service data
+- Uplink message
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/message-types/_index.md:15]: Okay yeah sorry about that comment about keeping this sorted. I didn't mean alphabetic but logical, so basically how it was.

File: doc/content/integrations/pubsub/message-types/_index.md [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+
+The Application Server can be configured to **subscribe** to messages to schedule the following events:
+- Downlink queue push
+- Downlink queue receive
```
Comment (johanstokking) [inline doc/content/integrations/pubsub/message-types/_index.md:19]: replace

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+In your application select the **Pub/Subs** submenu from the **Integrations** side menu. Clicking on the **+ Add Pub/Sub** button will open the Pub/Sub creation screen.
+
+Give your Pub/Sub an **ID**, choose a **Pub/Sub format**, and enter a **Base topic** to subscribe or publish to.
+
+Select **MQTT** as provider.
+
+{{< figure src="mqtt.png" alt="Pub/Sub creation screen" >}}
+
+If using client certificates, enable **Secure**. You will be asked to upload the **Root CA certificate**, the **Client certificate**, and the **Client private key.**
+
```
Comment (nejraselimovic) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:24]: I suggest rephrasing the first sentence into "If you want to use TLS for establishing secure connection with MQTT server, enable **Secure**". Also, I would shortly emphasize the location of generating those certificates and keys.

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 24]
```diff
@@ -0,0 +1,63 @@
...
+Enter the endpoint of the MQTT server in the **Server URL** field, including the protocol and port. 
+
```
Comment (nejraselimovic) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:26]: Here, I would point out that it is good to generally check the accessibility of the specified port on the given URL, before deploying the integration. Btw, is there a place a user can see all supported URL formats? I missed those in here.

File: doc/content/integrations/pubsub/mqtt-client/_index.md [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+Enter the **Client ID** the Application Server should use to authenticate.
```
Comment (nejraselimovic) [inline doc/content/integrations/pubsub/mqtt-client/_index.md:27]: Is the Application Server authenticating itself or the user?

File: doc/content/integrations/pubsub/mqtt-client/_index.md (hunk 24 above)
Comment (benolayinka) [reply doc/content/integrations/pubsub/mqtt-client/_index.md:26]: we should do a 'protocol' dropdown - making an issue for this

**Pull request merged**